### PR TITLE
=str #16787 java cookbook

### LINKED ISSUE #16787: Convert Cookbook to Java

### PR DESCRIPTION
**Pull request by 2m**
Resolves #16787

### CODE REVIEW

File: akka-docs-dev/rst/java/stream-cookbook.rst [hunk 1]
```diff
@@ -1,14 +1,386 @@
...
+
+Another approach to logging is to use ``log()`` operation which allows configuring logging for elements flowing through
+the stream as well as completion and erroring.
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeLoggingElements.java#log-custom
+
+Flattening a stream of sequences
+--------------------------------
+
+**Problem:** A stream is given as a stream of sequence of elements, but a stream of elements needed instead, streaming
```
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-cookbook.rst]: Seems we're using `Situation` everywhere, please change Problem to Situation :)

Comment (2m) [inline akka-docs-dev/rst/java/stream-cookbook.rst]: Nice catch.

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala [hunk 2]
```diff
@@ -22,16 +22,16 @@ class RecipeKeepAlive extends RecipeSpec {
...
 
       val graph = FlowGraph.closed() { implicit builder =>
         import FlowGraph.Implicits._
         val unfairMerge = builder.add(MergePreferred[ByteString](1))
 
-        dataStream ~> unfairMerge.preferred
         // If data is available then no keepalive is injected
-        keepAliveStream ~> unfairMerge ~> sink
+        dataStream ~> unfairMerge.preferred
+        ticks ~> keepAliveStream ~> unfairMerge ~> sink
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: good one

Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: I disagree, the keepAliveStream name exactly meant to mean that it is a stream of KeepAlive messages, i.e. it is a periodic source of ByteStrings that encode KeepAlive messages in the given format. As a Flow, this name looks wrong to me, it is now a stage that transforms ticks to KA messages, but no longer a stream of such messages.

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeMissedTicks.scala [hunk 3]
```diff
@@ -18,13 +18,12 @@ class RecipeMissedTicks extends RecipeSpec {
...
-      // tickStream is a Source[Tick]
-      val missedTicks: Source[Int, Unit] =
-        tickStream.conflate(seed = (_) => 0)(
+      val missedTicks: Flow[Tick, Int, Unit] =
+        Flow[Tick].conflate(seed = (_) => 0)(
           (missedTicks, tick) => missedTicks + 1)
       //#missed-ticks
 
-      missedTicks.to(sink).run()
+      tickStream.via(missedTicks).to(sink).run()
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeMissedTicks.scala]: here it is an improvement though.

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala (hunk 2 above)
Comment (2m) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: I changed it from source to flow so the type of it is not polluted with implementation/testing details, such as materialization value beeing `TestPublisher.Probe<Tick>`. This is for Java side, and I changed Scala side to be symmetrical.

Changing name to `keepAliveFlow` sounds good?

Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: It is a bit too generic for a name. Maybe tickToKeepAlivePacket?

**Comment by drewhk** [comment]
LGTM!, great work!

**Comment by ktoso** [comment]
I read through the examples and looks good I think. LGTM from my side :)

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala (hunk 2 above)
Comment (2m) [inline akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeKeepAlive.scala]: :+1:

**Comment by ktoso** [comment]
btw, awesome job, it's pretty tough to translate those :-)

**Comment by 2m** [comment]
You did most of the work. :) I just fixed compilation errors and rewrote couple of flow graphs.

**Pull request merged**